### DISCUSSION TITLE
How to Highlight Spaces in WinEdt

### DISCUSSION THREAD

**Discussion by saxyx** [WinEdt-Team/WinEdt-Support#66] [1.1 General Help and Discussion]
Sometimes, I need to check whether there are trailing spaces at the end of a line or how many spaces are between two characters. However, spaces blend into the background, and it seems there is no way to detect them except by selecting them with the mouse.

WinEdt uses the option `TAB_DRAW=1` to highlight the tabs, so it would be great if there were a similar option to highlight spaces.

**Comment by Alex-Simonic** [comment]
One thing that can already be done is to add space to the screen translation (in the TranslationTables.ini):

`  #32 -> "·" // Space to U+00B7
`
  
And comment it out to disable it. This could be automated by making a macro and a menu item that toggles this and reloads the translation tables.

You can see a discussion on which Unicode characters can be used to represent spaces (if U+00B7 is not your choice) on [stackexchange](https://ux.stackexchange.com/questions/91255/how-can-i-best-display-a-blank-space-character#:~:text=To%20represent%20a%20space%20bar%20of%20a%20keyboard%2C,%28%E2%B8%B1%29%20or%20U%2B00B7%20Middle%20dot%20%28%C2%B7%29%20as%20alternative.).

The graphic representation of a Tab is more complicated since its width varies with the position. But for a space, a simple screen translation should do in most cases...

**Reply by saxyx** [reply]
Thank you for your reply and the solution you provided. The screen translation approach works well for me. However, I’m not able to write a macro myself to automate the toggle. Could you please provide the code for implementing such a macro and any important considerations?

**Reply by Alex-Simonic** [reply]
Perhaps this will be of interest to others as well. Assuming that you have a local instance of TranslationTables.ini with the line containing the space translation (as described above) you can execute the following macro to toggle the comment in front of the space translation:

```
// Toggle Space Screen Translation...
// Must contain a space translation (possibly commented out):
//  #32 -> "·" // Space to U+00B7

IfFileExists('%b\ConfigEx\TranslationTables.ini',!'JMP(!"TABLE_EXISTS");');

Prompt('A Local instance of TranslationTables.ini does not exist');
Exit;

:TABLE_EXISTS:: ==================================================================

PushTagsandRegisters;

ReadFile('%b\ConfigEx\TranslationTables.ini',9);

FindInString("%!9","<//  #32 -\> ",1,2,11);
IfOK(!'Relax;',!'JMP(!"ENABLED");');
// Remove the comment //
SubstituteInString("%!9","<//  #32 -","  #32 -",11,9);
JMP(!"REWRITE");

:ENABLED:: =======================================================================

FindInString("%!9","<  #32 -\> ",1,2,11);
IfOK(!'Relax;',!'JMP(!"not_Found");');
// Insert the comment //
SubstituteInString("%!9","<  #32 -","//  #32 -",11,9);

:REWRITE:: =======================================================================

WriteFile('%b\ConfigEx\TranslationTables.ini','%!9');
LoadConfig('%b\ConfigEx\TranslationTables.ini');
PopTagsandRegisters;
Exit;

:not_Found:: =======================================================================

PopTagsandRegisters;
Prompt('Space Translation is not defined in TranslationTables.ini');

End;

```
You probably know how to add the menu item executing this macro script...

**Reply by saxyx** [reply]
Thank you so much for the macro script! It worked perfectly and resolved my problem. I appreciate your help and the detailed explanation. Thanks again for your support!

**Reply by saxyx** [reply]
I noticed that `#31 -> "¶"` allows WinEdt to display line breaks with the ¶ symbol, and in my .tex file, the ¶ symbol is colored `#008b8b`. Is there a way to change the color of the `·` symbol generated by `#32 -> "·"`?

**Reply by Alex-Simonic** [reply]
Putting the following in Private.ini works for me:

```
[FILTER_SETS*]

FILTER_SET="Visible Space"
  ENABLED=1
  MODE_FILTER=""
  SET=:[" "]
  BEFORE=""
  AFTER=""
  BOLN_NOT_OK=0
  EOLN_NOT_OK=0
  STRICT_PRIORITY=0
  PRIORITY=9
    TEXT_COLOR="fscHRT"
    DEFAULT_FONT=1
```

Priority and Color is up to you. It will not apply to the leading spaces (because they are excluded from highlighting) but works for the rest.

**Reply by saxyx** [reply]
Thank you very very much !
